### ISSUE TITLE
Remove or refactor RedeemScript class

### ISSUE THREAD

**Issue by afk11** [Bit-Wasp/bitcoin-php#102]
This class is oriented towards dealing with P2SH multisig scripts only. There needs to be a better way of dealing with P2SH, which is far more general than the current focus of class. This is probably connected to having a smart script solver. 

P2SH scripts are deferred output scripts. The redeemScript is provided in a redeeming transaction such that the Interpreter class can pop this blob of Script, and execute it as it would an output script. It does this only once the initial output script (the p2sh output funding script, OP_HASH160 <hash> OP_EQUAL) evaluates to true if provided the correct redeemScript for this output. 

I believe the RedeemScript class could probably be replaced with ScriptInterface, or an extended ScriptInterface which can classify scripts and return appropriately parsed structures. 

My worry here is that there may be a strong coupling between script types the library can classify, and it's ability to validate transactions. There is definitely such a coupling for signing transactions. 

Note, that the ability to parse an output script is vital to ones ability to sign. If you can't assess the requirements of an output script, you don't know what signatures to provide, and so on. Bitcoin Core avoids this problem by only dealing with standard script types. 

By having an extended ScriptInterface, we can take care to use only the simple ScriptInterface throughout validation code, and make use of the extended interface wherever signing is necessary. 

I believe a Solver should be able to identify signable inputs to start with, and eventually should be able to suggest the types of parameters which can be passed to complicated scripts.

**Comment by rubensayshi** [comment]
seems acceptible to not have support for every possible script, as long as we have isstandard + some more.

a more generic aproach would be good indeed, because then it's probably also easier to add more non standard scripts in the future.

**Comment by afk11** [comment]
https://github.com/Bit-Wasp/bitcoin-php/pull/172

I've removed RedeemScript, since it was purely m-of-n. We now pass around plain old `ScriptInterface`'s where the P2SH script is required. 

`ScriptInfoInterface` implementations exist for p2pk, p2pkh, multisig, and a 'scripthash' type, which wraps a p2sh script. The ScriptHash type only works with types classifiable by `OutputClassifier`, and where the P2SH script is one of those where a `ScriptInfoInterface` exists - ie, the three above. 

This only affects signing, which is great, and it allows for new script types to be supported reasonably easy (add the classifier check, the `ScriptInfoInterface`, and the extra line in ScriptHash for the type. 

I added P2shScriptFactory - accessible via `ScriptFactory::p2sh()` and exposes the same methods supported by ScriptHash.

**Comment by afk11** [comment]
ScriptFactory::p2sh() now returns a P2shScript, which has getOutputScript() and getAddress. To avoid confusion, it's the only script implementation to do so. It's not important to use this class anywhere, as only ScriptInterface is type hinted. It's main purpose is to have the above methods to avoid highly coupled boilerplate..

For signing, a ScriptInfoInterface is used, so no changes there. Closing this!
